crates/bevy_ui/src/layout/mod.rs

```diff
@@ -168,68 +165,83 @@ without UI components as a child of an entity with UI components, results may be
             ..default()
         };
 
-        let camera_root_node_map = self.camera_entity_to_taffy.entry(camera_id).or_default();
-        let existing_roots = self.camera_roots.entry(camera_id).or_default();
-        let mut new_roots = Vec::new();
         for entity in children {
-            let node = *self.entity_to_taffy.get(&entity).unwrap();
-            let root_node = existing_roots
-                .iter()
-                .find(|n| n.user_root_node == node)
-                .cloned()
-                .unwrap_or_else(|| {
-                    if let Some(previous_parent) = self.taffy.parent(node) {
-                        // remove the root node from the previous implicit node's children
-                        self.taffy.remove_child(previous_parent, node).unwrap();
-                    }
-
-                    let viewport_node = *camera_root_node_map
-                        .entry(entity)
-                        .or_insert_with(|| self.taffy.new_leaf(viewport_style.clone()).unwrap());
-                    self.taffy.add_child(viewport_node, node).unwrap();
+            let user_root_node = *self
+                .entity_to_taffy
+                .get(&entity)
+                .expect("root node was not inserted yet or was removed from entity_to_taffy map");
+
+            let implicit_viewport_node = *self
+                .camera_root_to_viewport_taffy
+                .entry((camera_id, entity))
+                .or_insert_with(|| self.taffy.new_leaf(viewport_style.clone()).unwrap());
+
+            if let Some(previous_parent) = self.taffy.parent(user_root_node) {
+                // remove the root node from the previous implicit node's children
+                self.taffy
+                    .remove_child(previous_parent, user_root_node)
+                    .unwrap();
```

Comment:
I think that moving this outside the original `unwrap_or_else` means it eagerly removes the child from its previous parent and then immediately applies the new (or same one) right after, regardless of whether it needs to. It might be negligible, but still worth leaving a note.